cpp/include/raft/core/buffer.hpp

```diff
@@ -0,0 +1,386 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+#include "raft/core/logger.hpp"
+#include <cstddef>
+#include <iterator>
+#include <memory>
+#include <raft/core/detail/buffer_utils/buffer_copy.hpp>
+#include <raft/core/detail/buffer_utils/non_owning_buffer.hpp>
+#include <raft/core/detail/buffer_utils/owning_buffer.hpp>
+#include <raft/core/detail/const_agnostic.hpp>
+#include <raft/core/device_support.hpp>
+#include <raft/core/device_type.hpp>
+#include <raft/core/exceptions.hpp>
+#include <raft/core/execution_device_id.hpp>
+#include <raft/core/execution_stream.hpp>
+#include <raft/core/memory_type.hpp>
+#include <stdint.h>
+#include <utility>
+#include <variant>
+
+namespace raft {
+/**
+ * @brief A container which may or may not own its own data on host or device
+ *
+ */
+using index_type = std::size_t;
+template <typename T>
+struct buffer {
+  using index_type = std::size_t;
```

Comment:
We should either make this part of the template or just explicitly use `std::size_t` where it appears.